File: src/managers/diagnosticsmanager.cpp
#include "diagnosticsmanager.h"
#include "settingsmanager.h"

#include <QClipboard>
#include <QDateTime>
#include <QDesktopServices>
#include <QDir>
#include <QFile>
#include <QGuiApplication>
#include <QLoggingCategory>
#include <QStandardPaths>
#include <QSysInfo>
#include <QUrl>

Q_LOGGING_CATEGORY(lcDiag, "octave.diagnostics")

#ifndef OCTAVE_VERSION
#define OCTAVE_VERSION "dev"
#endif

DiagnosticsManager::DiagnosticsManager(QObject *parent) : QObject(parent) {}

QString DiagnosticsManager::logDir() const
{
    return SettingsManager::getAppDataDir() + QStringLiteral("/logs");
}

QString DiagnosticsManager::appVersion() const
{
    return QStringLiteral(OCTAVE_VERSION);
}

QString DiagnosticsManager::deviceInfo() const
{
    return QStringLiteral("%1 %2 (%3, %4)")
        .arg(QSysInfo::prettyProductName(), QSysInfo::currentCpuArchitecture(),
             QSysInfo::machineHostName(), QSysInfo::kernelVersion());
}

bool DiagnosticsManager::canOpenFolder() const
{
#if defined(Q_OS_ANDROID) || defined(Q_OS_IOS)
    return false;
#else
    return true;
#endif
}

QStringList DiagnosticsManager::logFiles() const
{
    QDir dir(logDir());
    QStringList files = dir.entryList({QStringLiteral("octave*.log*")}, QDir::Files, QDir::Name);
    for (QString &f : files)
        f = dir.absoluteFilePath(f);
    return files;
}

static QString tailOf(const QString &path, int maxLines)
{
    QFile f(path);
    if (!f.open(QIODevice::ReadOnly | QIODevice::Text))
        return {};
    // Read the last 256 KB at most; enough for a few hundred lines
    const qint64 window = 256 * 1024;
    if (f.size() > window)
        f.seek(f.size() - window);
    const QString text = QString::fromUtf8(f.readAll());
    QStringList lines = text.split(QLatin1Char('\n'), Qt::SkipEmptyParts);
    if (lines.size() > maxLines)
        lines = lines.mid(lines.size() - maxLines);
    return lines.join(QLatin1Char('\n'));
}

QString DiagnosticsManager::recentLogLines(int maxLines) const
{
    const QString dir = logDir();
    const QString main = tailOf(dir + QStringLiteral("/octave-cpp.log"), maxLines);
    const QString errors = tailOf(dir + QStringLiteral("/octave-cpp-error.log"), maxLines / 4);
    QString out = main.isEmpty() ? QStringLiteral("(no log yet)") : main;
    if (!errors.isEmpty())
        out += QStringLiteral("\n\n---- errors ----\n") + errors;
    return out;
}

bool DiagnosticsManager::copyLogsToClipboard(int maxLines)
{
    QClipboard *cb = QGuiApplication::clipboard();
    if (!cb)
        return false;
    cb->setText(QStringLiteral("OCTAVE %1 (%2 backend) on %3\n%4\n\n%5")
                    .arg(appVersion(), backendName(), deviceInfo(),
                         QDateTime::currentDateTime().toString(Qt::ISODate), recentLogLines(maxLines)));
    qCInfo(lcDiag) << "Logs copied to clipboard";
    return true;
}

QString DiagnosticsManager::exportLogs()
{
    QString base = QStandardPaths::writableLocation(QStandardPaths::DownloadLocation);
    if (base.isEmpty())
        base = QStandardPaths::writableLocation(QStandardPaths::DocumentsLocation);
    const QString target = base + QStringLiteral("/OCTAVE-logs-")
                           + QDateTime::currentDateTime().toString(QStringLiteral("yyyyMMdd-HHmmss"));
    if (!QDir().mkpath(target)) {
        qCWarning(lcDiag) << "Could not create" << target;
        return {};
    }
    int copied = 0;
    for (const QString &src : logFiles()) {
        if (QFile::copy(src, target + QLatin1Char('/') + QFileInfo(src).fileName()))
            ++copied;
    }
    QFile info(target + QStringLiteral("/device-info.txt"));
    if (info.open(QIODevice::WriteOnly | QIODevice::Text)) {
        info.write(QStringLiteral("OCTAVE %1 (%2 backend)\n%3\nexported %4\n")
                       .arg(appVersion(), backendName(), deviceInfo(), QDateTime::currentDateTime().toString(Qt::ISODate))
                       .toUtf8());
    }
    qCInfo(lcDiag) << "Exported" << copied << "log files to" << target;
    m_lastExportPath = target;
    emit lastExportPathChanged();
    return target;
}

bool DiagnosticsManager::openLogFolder()
{
    if (!canOpenFolder())
        return false;
    return QDesktopServices::openUrl(QUrl::fromLocalFile(logDir()));
}

File: backend/diagnostics_manager.py
"""Field diagnostics: lets the user read and get the logs out of the head
unit without a laptop (Settings > About > Diagnostics). Mirrors
src/managers/diagnosticsmanager.{h,cpp}. The log files themselves come from
backend/logging_config.py (Python) / src/util/logger.cpp (C++)."""

import os
import platform
import shutil
from datetime import datetime

from PySide6.QtCore import QObject, Property, Signal, Slot, QStandardPaths, QUrl
from PySide6.QtGui import QDesktopServices, QGuiApplication

from backend.logging_config import get_logger
from backend.settings_manager import get_app_data_dir
from backend.version import __version__

logger = get_logger(__name__)


def _tail(path: str, max_lines: int) -> str:
    try:
        with open(path, "rb") as f:
            f.seek(0, os.SEEK_END)
            size = f.tell()
            window = 256 * 1024   # enough for a few hundred lines
            f.seek(max(0, size - window))
            text = f.read().decode("utf-8", errors="replace")
    except OSError:
        return ""
    lines = [ln for ln in text.split("\n") if ln]
    return "\n".join(lines[-max_lines:])


class DiagnosticsManager(QObject):
    lastExportPathChanged = Signal()

    def __init__(self, parent=None):
        super().__init__(parent)
        self._last_export_path = ""

    @Property(str, constant=True)
    def logDir(self) -> str:
        return os.path.join(get_app_data_dir(), "logs")

    @Property(str, constant=True)
    def appVersion(self) -> str:
        return __version__

    @Property(str, constant=True)
    def backendName(self) -> str:
        return "Python"

    @Property(str, constant=True)
    def deviceInfo(self) -> str:
        return f"{platform.system()} {platform.release()} {platform.machine()} ({platform.node()}, Python {platform.python_version()})"

    @Property(str, notify=lastExportPathChanged)
    def lastExportPath(self) -> str:
        """Where exportLogs() last copied the files ("" until it has run)"""
        return self._last_export_path

    @Property(bool, constant=True)
    def canOpenFolder(self) -> bool:
        """Desktop only: the OS file manager can open the log folder"""
        return True

    def _log_files(self) -> list:
        d = self.logDir
        try:
            names = sorted(n for n in os.listdir(d) if n.startswith("octave") and ".log" in n)
        except OSError:
            return []
        return [os.path.join(d, n) for n in names]

    @Slot(int, result=str)
    def recentLogLines(self, max_lines: int = 200) -> str:
        """Last max_lines lines of the main log followed by the error log"""
        d = self.logDir
        main = _tail(os.path.join(d, "octave.log"), max_lines)
        errors = _tail(os.path.join(d, "octave-error.log"), max(1, max_lines // 4))
        out = main or "(no log yet)"
        if errors:
            out += "\n\n---- errors ----\n" + errors
        return out

    @Slot(int, result=bool)
    def copyLogsToClipboard(self, max_lines: int = 400) -> bool:
        """recentLogLines() plus a header (version, device) onto the clipboard"""
        cb = QGuiApplication.clipboard()
        if cb is None:
            return False
        cb.setText(f"OCTAVE {self.appVersion} ({self.backendName} backend) on {self.deviceInfo}\n"
                   f"{datetime.now().isoformat(timespec='seconds')}\n\n{self.recentLogLines(max_lines)}")
        logger.info("Logs copied to clipboard")
        return True

    @Slot(result=str)
    def exportLogs(self) -> str:
        """Copy every log file (current + rotated) into
        <Downloads>/OCTAVE-logs-<timestamp>/ where any app can attach them.
        Returns the folder, "" on failure."""
        base = QStandardPaths.writableLocation(QStandardPaths.StandardLocation.DownloadLocation) \
            or QStandardPaths.writableLocation(QStandardPaths.StandardLocation.DocumentsLocation)
        target = os.path.join(base, "OCTAVE-logs-" + datetime.now().strftime("%Y%m%d-%H%M%S"))
        try:
            os.makedirs(target, exist_ok=True)
            copied = 0
            for src in self._log_files():
                shutil.copy2(src, target)
                copied += 1
            with open(os.path.join(target, "device-info.txt"), "w", encoding="utf-8") as f:
                f.write(f"OCTAVE {self.appVersion} ({self.backendName} backend)\n{self.deviceInfo}\n"
                        f"exported {datetime.now().isoformat(timespec='seconds')}\n")
        except OSError as e:
            logger.warning(f"Could not export logs to {target}: {e}")
            return ""
        logger.info(f"Exported {copied} log files to {target}")
        self._last_export_path = target
        self.lastExportPathChanged.emit()
        return target

    @Slot(result=bool)
    def openLogFolder(self) -> bool:
        """Desktop: open the log folder in the file manager"""
        return bool(QDesktopServices.openUrl(QUrl.fromLocalFile(self.logDir)))
